Add SM2 curve parameters

jforissier opened the pull request:
Adds the curve parameters for the Chinese SM2 ECC algorithms [1].

Link: [1] http://www.gmbz.org.cn/upload/2018-07-24/1532401863206085511.pdf
Signed-off-by: Jerome Forissier <jerome@forissier.org>

chris-jones-arm commented:
Hi @jforissier, thank you for your PR. As SM2 is a a widely used algorithm, support would be appreciated in mbed TLS.

Can I infer from [your comments](https://github.com/OP-TEE/optee_os/pull/4362#issuecomment-770819355) that you would be willing to contribute the whole curve algorithm over time as well as the current curve parameters? 

As support for SM2 is likely to be a large chunk of work it could be good to schedule this in with the team to ensure that you receive timely reviews. For the time being I will raise this with the team and we will try to keep you updated in progressing this further. Thank you for your understanding.

Note: this was attempted before at https://github.com/ARMmbed/mbedtls/pull/1620 however it is unclear to me why this did not progress further.

chris-jones-arm commented:
As this is just the curve parameters for SM2, this is being picked up as part of our standard review process and should be reviewed soon.

jforissier commented:
@chris-jones-arm thanks, and sorry for not coming back to you sooner. I have more code for SM2 DSA (Digital Signature Algorithm) which plugs into `mbedtls_ecdsa_sign()`/`mbedtls_ecdsa_verify()`, as well as a new function to help compute the SM2 DSA hash to be signed. The SM3 hash algorithm is also part of it as a dependency. In addition I have the SM4 cipher in various modes of operation. Whether we will upstream this and how is still being discussed internally in Huawei.

As for SM2 KEP (Key Exchange Protocol) and PKE (Public Key Encryption), I don't have MBed TLS implementations *per se* but as you noticed I have done them for OP-TEE so that could potentially be moved into MBed, too.

chris-jones-arm commented:
@jforissier That sounds great! Feel free to open more PR's for any additional code that you would like to upstream, contributions are always welcome.

chris-jones-arm commented:
Currently `check_names` is failing CI (see Travis CI results). Please fix this before we review.

jforissier commented:
> Currently `check_names` is failing CI (see Travis CI results). Please fix this before we review.

The error message is not super helpful... any link to documentation about this?

gilles-peskine-arm commented:
Sorry, the way `check-names` is written makes it hard to have meaningful error messages. I'm glad you found the right fix. You also need to run `tests/scripts/check-generated-files.sh -u` after changing `config.h`.

jforissier commented:
Update:
- Fixed check-names
- Added a few forgotten things (to-do check list near `mbedtls_ecp_group_id`)
- Decided to change naming slightly, "sm2p256v1" rather than plain "sm2", because OpenSSL uses these names.

chris-jones-arm commented:
The build is now passing on Travis (public CI) but is still failing on Jenkins (private CI). Here are the relevant logs:
```
******************************************
* Testing with only curve: MBEDTLS_ECP_DP_SM2P256V1_ENABLED
******************************************
  CC    aes.c
  CC    aesni.c
  CC    arc4.c
  CC    aria.c
  CC    asn1parse.c
  CC    asn1write.c
...
  CC    ssl_ticket.c
  CC    debug.c
  CC    ssl_ciphersuites.c
ecp_curves.c: In function 'mbedtls_ecp_group_load':
ecp_curves.c:878:21: error: implicit declaration of function 'LOAD_GROUP_A' [-Werror=implicit-function-declaration]
             return( LOAD_GROUP_A( sm2p256v1 ) );
                     ^
ecp_curves.c:878:35: error: 'sm2p256v1' undeclared (first use in this function)
             return( LOAD_GROUP_A( sm2p256v1 ) );
                                   ^
ecp_curves.c:878:35: note: each undeclared identifier is reported only once for each function it appears in
cc1: all warnings being treated as errors
Makefile:265: recipe for target 'ecp_curves.o' failed
make[1]: *** [ecp_curves.o] Error 1
make[1]: *** Waiting for unfinished jobs....
make: *** [lib] Error 2
Makefile:17: recipe for target 'lib' failed
Failed to build: only MBEDTLS_ECP_DP_SM2P256V1_ENABLED
^^^^test_depends_curves: test/build: curves.pl (gcc): tests/scripts/curves.pl -> 1^^^^
```

th0ma5b commented:
Dependency: #4492.

jforissier commented:
> Dependency: #4492.

How so? I agree that SM2 PKE/KEP/DSA need SM3, but this is just introducing the curve parameters.

yuhaoth reviewed (changes requested):
Please squash the commits and add test case for it.

My suggestion is one commits for SM2 and another one for the test case. Please also add [changelog entry](https://github.com/ARMmbed/mbedtls/blob/development/CONTRIBUTING.md#tests).

jforissier commented:
> Please squash the commits 

Done.

> and add test case for it. My suggestion is one commits for SM2 and another one for the test case

What kind of tests would be suitable? This is just the curve, not that actual SM2 encryption/key exchange/signature algorithms.

> Please also add [changelog entry](https://github.com/ARMmbed/mbedtls/blob/development/CONTRIBUTING.md#tests).

Done.

gilles-peskine-arm commented:
> What kind of tests would be suitable? This is just the curve, not that actual SM2 encryption/key exchange/signature algorithms.

I guess the kind of things that are in `test_suite_ecp`, whatever parts can be done with the part that's already implemented.

Also, perhaps some negative tests to check that if the application tries to use a feature that isn't implemented (perhaps automatically based on some decoding of metadata), it gets a proper error code.

yuhaoth commented:
> What kind of tests would be suitable? This is just the curve, not that actual SM2 encryption/key exchange/signature algorithms.

Add test vector into tests/suites/test_suite_ecp.data is enough

jforissier commented:
@gilles-peskine-arm @yuhaoth update

yuhaoth commented:
@jforissier , I think it is Okay for me , But there some conflicting files . Could you resolve that

jforissier commented:
> @jforissier , I think it is Okay for me , But there some conflicting files . Could you resolve that

Sure, done. Thanks for reviewing.

yuhaoth commented:
> 
> 
> > @jforissier , I think it is Okay for me , But there some conflicting files . Could you resolve that
> 
> Sure, done. Thanks for reviewing.

Thanks .

yuhaoth reviewed (approved):
I have checked the code and commit messages. 
It has include changelog entry and new test case

It looks good to me.

gilles-peskine-arm commented on library/ecp_curves.c:
```diff
@@ -580,6 +581,41 @@ static inline void ecp_mpi_load( mbedtls_mpi *X, const mbedtls_mpi_uint *p, size
 }
 
+/*
+ * Domain parameters for SM2 (GM/T 0003 Part 5)
```
Please put these together with the others, i.e. just before the definition of `ECP_LOAD_GROUP`.

jforissier replied:
OK

gilles-peskine-arm commented on library/ecp_curves.c:
```diff
@@ -580,6 +581,41 @@ static inline void ecp_mpi_load( mbedtls_mpi *X, const mbedtls_mpi_uint *p, size
+    BYTES_TO_T_UINT_8( 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF ),
+    BYTES_TO_T_UINT_8( 0xFF, 0xFF, 0xFF, 0xFF, 0xFE, 0xFF, 0xFF, 0xFF ),
+};
+#endif /* MBEDTLS_ECP_DP_BP512R1_ENABLED */
```
Copypasta

jforissier replied:
:facepalm:

gilles-peskine-arm commented on library/ecp_curves.c:
```diff
@@ -834,6 +870,11 @@ int mbedtls_ecp_group_load( mbedtls_ecp_group *grp, mbedtls_ecp_group_id id )
             return( LOAD_GROUP_A( brainpoolP384r1 ) );
 #endif /* MBEDTLS_ECP_DP_BP384R1_ENABLED */
 
+#if defined(MBEDTLS_ECP_DP_SM2P256V1_ENABLED)
```
Please put this after the BP512R1 case. Having a consistent order makes it easier to maintain the code.

jforissier replied:
OK

gilles-peskine-arm commented on library/ecp.c:
```diff
@@ -551,6 +551,10 @@ static const mbedtls_ecp_curve_info ecp_supported_curves[] =
     { MBEDTLS_ECP_DP_CURVE25519,   29,     256,    "x25519"            },
 #endif
+#if defined(MBEDTLS_ECP_DP_SM2P256V1_ENABLED)
+    /* https://tools.ietf.org/id/draft-yang-tls-tls13-sm-suites-05.html */
```
Should presumably be updated to [RFC 8998](https://datatracker.ietf.org/doc/html/rfc8998)

jforissier replied:
Indeed, thanks for the reference

gilles-peskine-arm reviewed (changes requested):
* What's there is mostly ok, but there are a few issues to resolve.
* I haven't checked that the numbers in the definition are correct.
* This PR enables SM2 for ECDSA and ECDH. We need unit tests for that, including some known-answer tests.
* RFC 8998 states that SM2 is only to be used for TLS 1.3, but this PR allows it (by default, even) in any version. Is there another official or draft standard that allows SM2 in (D)TLS 1.2?
* This PR enables SM2 in TLS by default. Assuming it's even ok to enable it in TLS 1.2, I don't like doing this without having at least interoperability tests.

jforissier commented:
>     * What's there is mostly ok, but there are a few issues to resolve.
> 
>     * I haven't checked that the numbers in the definition are correct.
> 
>     * This PR enables SM2 for ECDSA and ECDH. We need unit tests for that, including some known-answer tests.

It wasn't the intent. I am not sure it makes sense at all, because presumably this curve is to be used only with the custom signature and key exchange algorithms defined in the SM2 spec. Is there a way we could disable this curve in ECDSA and ECDH?

>     * RFC 8998 states that SM2 is only to be used for TLS 1.3, but this PR allows it (by default, even) in any version. Is there another official or draft standard that allows SM2 in (D)TLS 1.2?
> 
>     * This PR enables SM2 in TLS by default. Assuming it's even ok to enable it in TLS 1.2, I don't like doing this without having at least interoperability tests.

Here also, it wasn't the intent to extend TLS at this stage. Is there a way we could at least restrict SM2 to TLS 1.3? Would that even comply with the TLS 1.3 RFC? As mentioned above, I would think the custom SM2 algorithms are needed, not just using the SM2 curve with the existing/default algorithms.

This PR was meant as a first step towards adding some SM2 specific signature / key exchange / encryption functions to the MBed TLS API (and potentially using them for TLS of course).

I have addressed your other comments in an update commit.

gilles-peskine-arm commented:
At least, `mbedtls_ecdh_can_do` and `mbedtls_ecdsa_can_do` should return 0. I think this is enough to remove them from consideration in TLS, but I'd want to confirm this with a test. To exclude the curves from explicit requests to do ECDH/ECDSA, I think explicit checks in the ECDH and ECDSA operation code is needed.

jforissier commented:
> At least, `mbedtls_ecdh_can_do` and `mbedtls_ecdsa_can_do` should return 0.

Done.

> I think this is enough to remove them from consideration in TLS, but I'd want to confirm this with a test.

I can add a test if you guide me where I should poke ;)

> To exclude the curves from explicit requests to do ECDH/ECDSA, I think explicit checks in the ECDH and ECDSA operation code is needed.

ECDH already calls mbedtls_ecdh_can_do() in ecdh_compute_shared_restartable(), but ECDSA doesn't seem to check curve support so I have added a call to mbedtls_ecdsa_can_do() in ecdh_compute_shared_restartable().

jforissier commented:
Anticipating on the resolution of the merge conflict... how do you expect contributors of new curves will figure out what to add to support `MBEDTLS_ECP_FIXED_POINT_OPTIM == 1` as modified by ba9cb76e9? :confused:

gilles-peskine-arm commented:
> I can add a test if you guide me where I should poke ;)

One possible way: in `ssl-opt.sh`: passing `force_ciphersuite= curve_list=sm2,secp256r1` to both the client and the server should result in logs showing that secp256r1 is used for both ECDH and ECDSA.

Another possible way: leverage `perform_handshake` in `test_suite_ssl.function`. I don't know how easy it is to adapt it for a new way in which the handshake is expected to fail.

gilles-peskine-arm commented:
> how do you expect contributors of new curves will figure out what to add to support MBEDTLS_ECP_FIXED_POINT_OPTIM == 1 as modified by ba9cb76? 

I think you have to understand the arithmetic better than I do. @mpg ?

mpg commented:
> I think you have to understand the arithmetic better than I do. 

I don't think that should be needed. There's a script for that, with usage instructions: https://github.com/ARMmbed/mbedtls/blob/development/scripts/ecp_comb_table.py#L29 If the instructions are not sufficient (and this PR is the ideal occasion to test that), we should improve them.

Apparently the script is not be discoverable enough either, so perhaps we should mention it in strategic places in the library source too. Perhaps for each curve's `T` table: "This was generated with scripts/ecp_comb_table.py" - hopefully each person who adds a new curve will have a look at at least one of the existing curves. @jforissier wdyt?

(Note that a nice aspect of #4315 is that new curves will work even without a pre-computed table - they just won't have any performance boost for `FIXED_POINT_OPTIM == 1` until the embedded table is added, but will still work.)

jforissier commented:
Rebased with comb table issue addressed. I'm still looking at how to add the TLS negative test(s).

yuhaoth commented:
https://travis-ci.com/github/ARMmbed/mbedtls/jobs/512019101#L2106 ， test fail . Please fix that

jforissier commented:
> https://travis-ci.com/github/ARMmbed/mbedtls/jobs/512019101#L2106 ， test fail . Please fix that

It passes here:
```
DHM size: server default, client 2049, rejected ........................ Terminated
PASS
DHM: no SM2 ............................................................ Terminated
PASS
PSK callback: psk, no callback ......................................... Terminated
PASS
```

But many other tests in `tests/ssl-opt.sh` fail in my environment. Any idea what's going on? Different versions of `openssl` perhaps?

gilles-peskine-arm commented:
The SSL testing scripts need specific builds of openssl and gnutls. See https://developer.trustedfirmware.org/w/mbed-tls/testing/ci/#tooling-for-all-sh

jforissier commented:
> The SSL testing scripts need specific builds of openssl and gnutls. See https://developer.trustedfirmware.org/w/mbed-tls/testing/ci/#tooling-for-all-sh

:open_mouth: honestly this is a bit too much... Unless you can give me a simple command to have everything in place on Ubuntu 21.04, I will simply drop this PR. It shouldn't be more complicated that clone + make check really.
:cry:

gilles-peskine-arm commented:
> It shouldn't be more complicated that clone + make check really.

That will give you tests in ONE configuration. Mbed TLS is primarily intended for embedded devices and has many compile-time configuration options, so you can't test it from just one build.

> Unless you can give me a simple command to have everything in place on Ubuntu 21.04

Try the [Dockerfile](https://github.com/ARMmbed/mbedtls/blob/master/tests/docker/bionic/Dockerfile).

mpg commented:
One alternative is to use the `-f` or `-e` options of `ssl-opt.sh` to run only the tests you're interested it, and don't worry about those that require a specific version of OpenSSL or GnuTLS. For example `tests/ssl-opt.sh -e 'gnutls\|openssl'` should run all tests except those that require OpenSSL or Gnu TLS.

Though to be honest, I rarely if ever run the whole script on my machine, in this case I would use `tests/ssl-opt.sh -f SM2`. 

I think Gilles' comment on testing multiple configurations was spot on: on my machine `tests/ssl-opt.sh -f SM2` passes if I build with the default configuration, but fails with the same error as on Travis if I run `scripts/config.py full` before building (which is what this test on Travis does). (Note : our Makefile doesn't have proper dependency tracking, so you'll want to run `make clean` after invoking `scripts/config.py` or otherwise modifying the `config.h` file.)

mpg commented on tests/ssl-opt.sh:
```diff
@@ -5221,6 +5221,18 @@ run_test    "DHM size: server default, client 2049, rejected" \
             -c "DHM prime too short:"
 
+# Check that the SM2 curve is not used for DHM
+run_test    "DHM: no SM2" \
```
Minor: by DHM we usually mean finite-field Diffie-Hellman(-Merkle), please use ECDH here and in the comment above.

jforissier commented:
> I think Gilles' comment on testing multiple configurations was spot on: on my machine `tests/ssl-opt.sh -f SM2` passes if I build with the default configuration, but fails with the same error as on Travis if I run `scripts/config.py full` before building (which is what this test on Travis does). (Note : our Makefile doesn't have proper dependency tracking, so you'll want to run `make clean` after invoking `scripts/config.py` or otherwise modifying the `config.h` file.)

That was quite helpful. I used `scripts/config.py full` and `git diff`, then enabled features one by one with `scripts/config.py set <FEATURE>` followed by `make clean && make -j10 && make -j10 programs && ./tests/ssl-opt.sh -f SM2` until I found that `MBEDTLS_USE_PSA_CRYPTO` was the key thing making the test fail.

PR updated.

gilles-peskine-arm commented:
There are merge conflicts again: `config.h` has changed and has been renamed, and a trivial conflict due to co-located changes in a test data file. Now that Mbed TLS 3.0 has been released, I don't expect conflicts to arise as often. Please rebase.

tom-daubney-arm commented:
We are now converting older PRs to draft PRs where the following conditions are met: They have not been updated in the last 3 months, and they need more than non-trivial work to complete.
